File: BlinkID/android/src/main/java/com/microblink/flutter/MicroblinkScanner.java
package com.microblink.flutter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

import com.microblink.MicroblinkSDK;
import com.microblink.directApi.DirectApiErrorListener;
import com.microblink.directApi.RecognizerRunner;
import com.microblink.entities.recognizers.RecognizerBundle;
import com.microblink.flutter.recognizers.RecognizerSerializers;
import com.microblink.hardware.orientation.Orientation;
import com.microblink.intent.IntentDataTransferMode;
import com.microblink.metadata.MetadataCallbacks;
import com.microblink.metadata.detection.quad.DisplayableQuadDetection;
import com.microblink.metadata.detection.quad.QuadDetectionCallback;
import com.microblink.metadata.recognition.FirstSideRecognitionCallback;
import com.microblink.recognition.RecognitionSuccessType;
import com.microblink.view.recognition.DetectionStatus;
import com.microblink.view.recognition.ScanResultListener;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Map;

public class MicroblinkScanner implements ImageAnalysis.Analyzer {
    private final RecognizerRunner recognizerRunner;
    private final RecognizerBundle recognizerBundle;
    private final MicroblinkScanner.Callbacks callbacks;

    MicroblinkScanner(
            Context context,
            MicroblinkCreationParams creationParams,
            MicroblinkScanner.Callbacks callbacks
    ) {
        setLicense(context, creationParams.licenseKey);

        this.callbacks = callbacks;
        recognizerBundle = createRecognizerBundle((Map<String, Object>) creationParams.recognizerCollection);
        recognizerRunner = createRecognizer(context, recognizerBundle);
        recognizerRunner.setMetadataCallbacks(createMetadataCallbacks(callbacks));
    }

    @SuppressLint("NewApi")
    @Override
    public void analyze(@NonNull ImageProxy image) {
        if (recognizerRunner.getCurrentState() == RecognizerRunner.State.READY) {
            byte[] rgbaBytes = new byte[image.getPlanes()[0].getBuffer().limit()];
            image.getPlanes()[0].getBuffer().get(rgbaBytes);
            Bitmap bitmap = bitmapFromRgba(rgbaBytes, image.getWidth(), image.getHeight());

            recognizerRunner.recognizeBitmap(bitmap, Orientation.ORIENTATION_LANDSCAPE_RIGHT,
                    createScanResultListener(image));
        } else {
            image.close();
        }
    }

    private ScanResultListener createScanResultListener(final ImageProxy imageProxy) {
        return new ScanResultListener() {
            @SuppressLint("NewApi")
            @Override
            public void onScanningDone(@NonNull RecognitionSuccessType recognitionSuccessType) {
                JSONArray resultJsonArray =
                        RecognizerSerializers.INSTANCE.serializeRecognizerResults(
                                recognizerBundle.getRecognizers());
                callbacks.onScanned(resultJsonArray.toString());
                imageProxy.close();
            }

            @Override
            public void onUnrecoverableError(@NonNull Throwable throwable) {
                callbacks.onError(throwable);
                recognizerRunner.resetRecognitionState();
            }
        };
    }

    private Bitmap bitmapFromRgba(byte[] bytes, int width, int height) {
        int[] colorPixels = new int[width * height];
        int j = 0;

        for (int i = 0; i < colorPixels.length; i++) {
            byte R = bytes[j++];
            byte G = bytes[j++];
            byte B = bytes[j++];
            byte A = bytes[j++];

            colorPixels[i] = (A & 0xff) << 24 | (R & 0xff) << 16 | (G & 0xff) << 8 | (B & 0xff);
        }

        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        bitmap.setPixels(colorPixels, 0, width, 0, 0, width, height);
        return bitmap;
    }

    private void setLicense(Context context, String licenseKey) {
        MicroblinkSDK.setShowTrialLicenseWarning(false);
        MicroblinkSDK.setLicenseKey(licenseKey, context);
        MicroblinkSDK.setIntentDataTransferMode(IntentDataTransferMode.PERSISTED_OPTIMISED);
    }

    private RecognizerBundle createRecognizerBundle(Map<String, Object> recognizerCollection) {
        JSONObject jsonRecognizerCollection = new JSONObject(
                recognizerCollection);

        return RecognizerSerializers.INSTANCE.deserializeRecognizerCollection(
                jsonRecognizerCollection);
    }

    private RecognizerRunner createRecognizer(Context context, RecognizerBundle recognizerBundle) {
        RecognizerRunner recognizerRunner = RecognizerRunner.getSingletonInstance();

        recognizerRunner.initialize(context, recognizerBundle, new DirectApiErrorListener() {
            @Override
            public void onRecognizerError(@NonNull Throwable throwable) {
                callbacks.onError(throwable);
            }
        });

        return recognizerRunner;
    }

    private MetadataCallbacks createMetadataCallbacks(final MicroblinkScanner.Callbacks callbacks) {
        MetadataCallbacks metadataCallbacks = new MetadataCallbacks();

        metadataCallbacks.setQuadDetectionCallback(new QuadDetectionCallback() {
            @Override
            public void onQuadDetection(
                    final @NonNull DisplayableQuadDetection displayableQuadDetection) {
                callbacks.onDetectionStatusUpdated(displayableQuadDetection.getDetectionStatus());
            }
        });

        metadataCallbacks.setFirstSideRecognitionCallback(new FirstSideRecognitionCallback() {
            @Override
            public void onFirstSideRecognitionFinished() {
                callbacks.onFirstSideScanned();
            }
        });

        return metadataCallbacks;
    }

    public void dispose() {
        recognizerRunner.terminate();
    }

    interface Callbacks {
        void onFirstSideScanned();

        void onDetectionStatusUpdated(DetectionStatus detectionStatus);

        void onScanned(String result);

        void onError(Throwable throwable);
    }
}

File: BlinkID/android/src/main/kotlin/com/microblink/blinkid/flutter/MicroblinkScanner.kt
package com.microblink.blinkid.flutter

import android.content.Context
import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.microblink.blinkid.MicroblinkSDK
import com.microblink.blinkid.directApi.RecognizerRunner
import com.microblink.blinkid.entities.recognizers.Recognizer
import com.microblink.blinkid.entities.recognizers.RecognizerBundle
import com.microblink.blinkid.flutter.recognizers.RecognizerSerializers
import com.microblink.blinkid.hardware.orientation.Orientation
import com.microblink.blinkid.image.Image
import com.microblink.blinkid.image.ImageBuilder
import com.microblink.blinkid.image.InputImage
import com.microblink.blinkid.intent.IntentDataTransferMode
import com.microblink.blinkid.metadata.MetadataCallbacks
import com.microblink.blinkid.metadata.detection.quad.DisplayableQuadDetection
import com.microblink.blinkid.metadata.detection.quad.QuadDetectionCallback
import com.microblink.blinkid.metadata.recognition.FirstSideRecognitionCallback
import com.microblink.blinkid.recognition.RecognitionSuccessType
import com.microblink.blinkid.view.recognition.ScanResultListener
import org.json.JSONObject

class MicroblinkScanner internal constructor(
    context: Context,
    creationParams: MicroblinkCreationParams,
    callbacks: Callbacks
) : ImageAnalysis.Analyzer {
    private val recognizerRunner: RecognizerRunner
    private val recognizerBundle: RecognizerBundle
    private val callbacks: Callbacks
    private var isPaused = false

    init {
        com.microblink.blinkid.util.Log.setLogLevel(com.microblink.blinkid.util.Log.LogLevel.LOG_VERBOSE);
        setLicense(context, creationParams.licenseKey)
        this.callbacks = callbacks
        recognizerBundle = createRecognizerBundle(creationParams.recognizerCollection)
        recognizerRunner = createRecognizer(context, recognizerBundle)
        recognizerRunner.setMetadataCallbacks(createMetadataCallbacks(callbacks))
    }

    @OptIn(ExperimentalGetImage::class)
    override fun analyze(imageProxy: ImageProxy) {
        if (!isPaused && recognizerRunner.currentState == RecognizerRunner.State.READY) {
            imageProxy.image?.let {
                val image = ImageBuilder.buildInputImageFromCamera2Image(it, Orientation.ORIENTATION_LANDSCAPE_RIGHT, null)
                recognizerRunner.recognizeVideoImage(image, createScanResultListener(image, imageProxy))
            }
        } else {
            imageProxy.close()
        }
    }

    private fun createScanResultListener(image: InputImage, imageProxy: ImageProxy): ScanResultListener {
        return object : ScanResultListener {
            override fun onScanningDone(recognitionSuccessType: RecognitionSuccessType) {
                imageProxy.close()
                if (recognitionSuccessType == RecognitionSuccessType.UNSUCCESSFUL) {
                    return
                }

                val recognizers = recognizerBundle.recognizers
                recognizers.forEach { recognizer ->
                    val resultState = recognizer.result.resultState
                    callbacks.onScanningDone(resultState)
                    if (resultState == Recognizer.Result.State.Valid) {
                        isPaused = true
                        callbacks.onScanned(recognizers.clone())
                        return
                    }
                }
            }

            override fun onUnrecoverableError(throwable: Throwable) {
                imageProxy.close()
                callbacks.onError(throwable)
                recognizerRunner.resetRecognitionState()
            }
        }
    }

    fun resume() {
        recognizerRunner.resetRecognitionState()
        isPaused = false
    }

    private fun setLicense(context: Context, licenseKey: String?) {
        MicroblinkSDK.setShowTrialLicenseWarning(false)
        MicroblinkSDK.setLicenseKey(licenseKey!!, context)
        MicroblinkSDK.setIntentDataTransferMode(IntentDataTransferMode.OPTIMISED)
    }

    private fun createRecognizerBundle(recognizerCollection: Map<String?, Any?>): RecognizerBundle {
        val jsonRecognizerCollection = JSONObject(recognizerCollection)
        return RecognizerSerializers.INSTANCE.deserializeRecognizerCollection(jsonRecognizerCollection)
    }

    private fun createRecognizer(context: Context, recognizerBundle: RecognizerBundle): RecognizerRunner {
        val recognizerRunner = RecognizerRunner.getSingletonInstance()
        recognizerRunner.initialize(context, recognizerBundle) { throwable -> callbacks.onError(throwable) }
        return recognizerRunner
    }

    private fun createMetadataCallbacks(callbacks: Callbacks): MetadataCallbacks {
        val metadataCallbacks = MetadataCallbacks()
        metadataCallbacks.quadDetectionCallback = QuadDetectionCallback { displayableQuadDetection ->
            callbacks.onDetectionStatusUpdated(displayableQuadDetection)
        }
        metadataCallbacks.firstSideRecognitionCallback = FirstSideRecognitionCallback { callbacks.onFirstSideScanned() }
        return metadataCallbacks
    }

    fun dispose() {
        recognizerRunner.terminate()
    }

    internal interface Callbacks {
        fun onFirstSideScanned()
        fun onDetectionStatusUpdated(displayableQuadDetection: DisplayableQuadDetection)
        fun onScanningDone(state: Recognizer.Result.State)
        fun onScanned(recognizers: Array<Recognizer<Recognizer.Result>>)
        fun onError(throwable: Throwable)
    }
}